File: tests/wasm/test_linreg.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('LinReg partial params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.linreg_js(close, 14);
    assert.strictEqual(result.length, close.length);
});

test('LinReg accuracy', async () => {

    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.linreg;

    const result = wasm.linreg_js(close, expected.defaultParams.period);

    assert.strictEqual(result.length, close.length);


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        0.1,
        "LinReg last 5 values mismatch"
    );


    await compareWithRust('linreg', result, 'close', expected.defaultParams);
});

test('LinReg default candles', async () => {

    const close = new Float64Array(testData.close);

    const result = wasm.linreg_js(close, 14);
    assert.strictEqual(result.length, close.length);


    await compareWithRust('linreg', result, 'close', { period: 14 });
});

test('LinReg zero period', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.linreg_js(inputData, 0);
    }, /Invalid period/);
});

test('LinReg period exceeds length', () => {

    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.linreg_js(dataSmall, 10);
    }, /Invalid period|Not enough valid data/);
});

test('LinReg very small dataset', () => {

    const dataSingle = new Float64Array([42.0]);

    assert.throws(() => {
        wasm.linreg_js(dataSingle, 14);
    }, /Invalid period|Not enough valid data/);
});

test('LinReg empty input', () => {

    const dataEmpty = new Float64Array([]);

    assert.throws(() => {
        wasm.linreg_js(dataEmpty, 14);
    }, /no data provided|empty|all values are nan/i);
});

test('LinReg all NaN', () => {

    const data = new Float64Array([NaN, NaN, NaN, NaN, NaN]);

    assert.throws(() => {
        wasm.linreg_js(data, 3);
    }, /All values are NaN/);
});

test('LinReg reinput', () => {

    const close = new Float64Array(testData.close);


    const firstResult = wasm.linreg_js(close, 14);


    const secondResult = wasm.linreg_js(firstResult, 10);

    assert.strictEqual(secondResult.length, firstResult.length);





    for (let i = 24; i < secondResult.length; i++) {
        assert(isFinite(secondResult[i]), `NaN found at index ${i}`);
    }
});

test('LinReg NaN handling', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.linreg_js(close, 14);

    assert.strictEqual(result.length, close.length);



    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
});

test('LinReg batch', () => {

    const close = new Float64Array(testData.close);


    const batchResult = wasm.linreg_batch(close, {
        period_range: [10, 40, 10]
    });


    assert(batchResult.values, 'Should have values array');
    assert(batchResult.combos, 'Should have combos array');
    assert(typeof batchResult.rows === 'number', 'Should have rows count');
    assert(typeof batchResult.cols === 'number', 'Should have cols count');


    assert.strictEqual(batchResult.rows, 4);
    assert.strictEqual(batchResult.cols, close.length);
    assert.strictEqual(batchResult.combos.length, 4);
    assert.strictEqual(batchResult.values.length, 4 * close.length);


    const expectedPeriods = [10, 20, 30, 40];
    for (let i = 0; i < 4; i++) {
        assert.strictEqual(batchResult.combos[i].period, expectedPeriods[i]);
    }


    for (let i = 0; i < 4; i++) {
        const period = expectedPeriods[i];
        const individual_result = wasm.linreg_js(close, period);


        const row_start = i * close.length;
        const row = batchResult.values.slice(row_start, row_start + close.length);

        assertArrayClose(row, individual_result, 1e-9, `Period ${period}`);
    }
});

test('LinReg different periods', () => {

    const close = new Float64Array(testData.close);


    for (const period of [5, 10, 20, 50]) {
        const result = wasm.linreg_js(close, period);
        assert.strictEqual(result.length, close.length);



        for (let i = 0; i < period - 1; i++) {
            assert(isNaN(result[i]), `Expected NaN at index ${i} for period=${period}`);
        }


        for (let i = period - 1; i < result.length; i++) {
            assert(!isNaN(result[i]), `Unexpected NaN at index ${i} for period=${period}`);
        }
    }
});

test('LinReg batch performance', () => {

    const close = new Float64Array(testData.close.slice(0, 1000));


    const startBatch = performance.now();
    const batchResult = wasm.linreg_batch(close, {
        period_range: [10, 50, 10]
    });
    const batchTime = performance.now() - startBatch;

    const startSingle = performance.now();
    const singleResults = [];
    for (let period = 10; period <= 50; period += 10) {
        singleResults.push(...wasm.linreg_js(close, period));
    }
    const singleTime = performance.now() - startSingle;


    console.log(`Batch time: ${batchTime.toFixed(2)}ms, Single time: ${singleTime.toFixed(2)}ms`);


    assertArrayClose(batchResult.values, singleResults, 1e-9, 'Batch vs single results');
});

test('LinReg edge cases', () => {



    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0]);
    const result = wasm.linreg_js(data, 3);
    assert.strictEqual(result.length, data.length);



    assertClose(result[result.length - 1], 10.0, 1e-9, "Perfect linear regression failed");


    const constantData = new Float64Array(20).fill(5.0);
    const constantResult = wasm.linreg_js(constantData, 5);
    assert.strictEqual(constantResult.length, constantData.length);


    for (let i = 5; i < constantResult.length; i++) {
        assertClose(constantResult[i], 5.0, 1e-9, `Constant prediction failed at index ${i}`);
    }
});

test('LinReg single value', () => {

    const data = new Float64Array([42.0]);


    const result = wasm.linreg_js(data, 1);
    assert.strictEqual(result.length, 1);
    assert(isNaN(result[0]));
});

test('LinReg two values', () => {

    const data = new Float64Array([1.0, 2.0]);



    const result = wasm.linreg_js(data, 1);
    assert.strictEqual(result.length, 2);
    assert(isNaN(result[0]));
    assert(isNaN(result[1]));


    const result2 = wasm.linreg_js(data, 2);
    assert.strictEqual(result2.length, 2);
    assert(isNaN(result2[0]));
    assertClose(result2[1], 2.0, 1e-9, "Two-value prediction failed");
});

test('LinReg batch metadata', () => {

    const close = new Float64Array(50);
    close.fill(100);

    const result = wasm.linreg_batch(close, {
        period_range: [15, 45, 15]
    });


    assert.strictEqual(result.combos.length, 3);
    assert.strictEqual(result.combos[0].period, 15);
    assert.strictEqual(result.combos[1].period, 30);
    assert.strictEqual(result.combos[2].period, 45);
});

test('LinReg warmup period calculation', () => {

    const close = new Float64Array(testData.close.slice(0, 50));

    const testCases = [
        { period: 5, expectedWarmup: 5 },
        { period: 10, expectedWarmup: 10 },
        { period: 20, expectedWarmup: 20 },
        { period: 30, expectedWarmup: 30 },
    ];

    for (const { period, expectedWarmup } of testCases) {
        const result = wasm.linreg_js(close, period);



        for (let i = 0; i < period - 1 && i < result.length; i++) {
            assert(isNaN(result[i]), `Expected NaN at index ${i} for period=${period}`);
        }


        if (period - 1 < result.length) {
            assert(!isNaN(result[period - 1]),
                `Expected valid value at index ${period - 1} for period=${period}`);
        }
    }
});

test('LinReg consistency across calls', () => {

    const close = new Float64Array(testData.close.slice(0, 100));

    const result1 = wasm.linreg_js(close, 14);
    const result2 = wasm.linreg_js(close, 14);

    assertArrayClose(result1, result2, 1e-15, "LinReg results not consistent");
});

test('LinReg parameter step precision', () => {

    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);

    const batchResult = wasm.linreg_batch(data, {
        period_range: [2, 4, 1]
    });


    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.combos.length, 3);
    assert.strictEqual(batchResult.values.length, 3 * data.length);


    assert.strictEqual(batchResult.combos[0].period, 2);
    assert.strictEqual(batchResult.combos[1].period, 3);
    assert.strictEqual(batchResult.combos[2].period, 4);
});

test('LinReg slope calculation', () => {


    const data = new Float64Array([2.0, 4.0, 6.0, 8.0, 10.0, 12.0, 14.0, 16.0]);
    const result = wasm.linreg_js(data, 4);



    assertClose(result[result.length - 1], 16.0, 1e-9, "Slope calculation failed");
});

test('LinReg streaming simulation', () => {

    const close = new Float64Array(testData.close.slice(0, 100));
    const period = 14;


    const batchResult = wasm.linreg_js(close, period);



    assert.strictEqual(batchResult.length, close.length);



    for (let i = 0; i < period - 1; i++) {
        assert(isNaN(batchResult[i]), `Expected NaN at index ${i}`);
    }


    for (let i = period - 1; i < close.length; i++) {
        assert(isFinite(batchResult[i]), `Expected finite value at index ${i}`);
    }
});

test('LinReg large period', () => {

    const data = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        data[i] = Math.sin(i * 0.1) * 100 + 1000;
    }

    const result = wasm.linreg_js(data, 99);
    assert.strictEqual(result.length, data.length);


    for (let i = 0; i < 98; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i}`);
    }


    assert(isFinite(result[98]), "Expected finite value at index 98");
    assert(isFinite(result[99]), "Expected finite value at last index");
});



test('LinReg zero-copy basic', () => {

    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    const period = 3;


    const ptr = wasm.linreg_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');


    const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, data.length);


    memView.set(data);


    try {
        wasm.linreg_into(ptr, ptr, data.length, period);


        const regularResult = wasm.linreg_js(data, period);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue;
            }
            assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
        }
    } finally {

        wasm.linreg_free(ptr, data.length);
    }
});

test('LinReg zero-copy with large dataset', () => {
    const size = 100000;
    const data = new Float64Array(size);
    for (let i = 0; i < size; i++) {
        data[i] = Math.sin(i * 0.01) + Math.random() * 0.1;
    }

    const ptr = wasm.linreg_alloc(size);
    assert(ptr !== 0, 'Failed to allocate large buffer');

    try {
        const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        memView.set(data);

        wasm.linreg_into(ptr, ptr, size, 14);


        const memView2 = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);


        for (let i = 0; i < 13; i++) {
            assert(isNaN(memView2[i]), `Expected NaN at warmup index ${i}`);
        }


        for (let i = 13; i < Math.min(100, size); i++) {
            assert(!isNaN(memView2[i]), `Unexpected NaN at index ${i}`);
        }
    } finally {
        wasm.linreg_free(ptr, size);
    }
});

test('LinReg zero-copy error handling', () => {

    assert.throws(() => {
        wasm.linreg_into(0, 0, 10, 5);
    }, /null pointer/);


    const ptr = wasm.linreg_alloc(10);
    try {
        assert.throws(() => {
            wasm.linreg_into(ptr, ptr, 10, 0);
        }, /Invalid period/);
    } finally {
        wasm.linreg_free(ptr, 10);
    }
});

test('LinReg zero-copy batch API', () => {
    const size = 100;
    const data = new Float64Array(size);
    for (let i = 0; i < size; i++) {
        data[i] = Math.sin(i * 0.1) * 10;
    }


    const periods = 3;
    const totalSize = periods * size;
    const inPtr = wasm.linreg_alloc(size);
    const outPtr = wasm.linreg_alloc(totalSize);

    try {
        const inView = new Float64Array(wasm.__wasm.memory.buffer, inPtr, size);
        inView.set(data);

        const rows = wasm.linreg_batch_into(inPtr, outPtr, size, 10, 30, 10);
        assert.strictEqual(rows, 3, 'Expected 3 rows');

        const outView = new Float64Array(wasm.__wasm.memory.buffer, outPtr, totalSize);


        const periodValues = [10, 20, 30];
        for (let i = 0; i < periods; i++) {
            const period = periodValues[i];
            const individual = wasm.linreg_js(data, period);
            const rowStart = i * size;
            const row = Array.from(outView.slice(rowStart, rowStart + size));

            assertArrayClose(row, individual, 1e-9, `Batch row ${i} mismatch`);
        }
    } finally {
        wasm.linreg_free(inPtr, size);
        wasm.linreg_free(outPtr, totalSize);
    }
});



test('LinReg SIMD consistency', () => {


    const testCases = [
        { size: 10, period: 5 },
        { size: 100, period: 14 },
        { size: 1000, period: 20 },
        { size: 10000, period: 50 }
    ];

    for (const testCase of testCases) {
        const data = new Float64Array(testCase.size);
        for (let i = 0; i < testCase.size; i++) {
            data[i] = Math.sin(i * 0.1) + Math.cos(i * 0.05);
        }

        const result = wasm.linreg_js(data, testCase.period);


        assert.strictEqual(result.length, data.length);


        for (let i = 0; i < testCase.period - 1; i++) {
            assert(isNaN(result[i]), `Expected NaN at warmup index ${i} for size=${testCase.size}`);
        }


        let sumAfterWarmup = 0;
        let countAfterWarmup = 0;
        for (let i = testCase.period - 1; i < result.length; i++) {
            assert(!isNaN(result[i]), `Unexpected NaN at index ${i} for size=${testCase.size}`);
            sumAfterWarmup += result[i];
            countAfterWarmup++;
        }


        const avgAfterWarmup = sumAfterWarmup / countAfterWarmup;
        assert(Math.abs(avgAfterWarmup) < 1000, `Average value ${avgAfterWarmup} seems unreasonable`);
    }
});



test('LinReg batch metadata validation', () => {

    const close = new Float64Array(50);
    close.fill(100);

    const result = wasm.linreg_batch(close, {
        period_range: [10, 30, 10]
    });


    assert.strictEqual(result.combos.length, 3);
    assert.strictEqual(result.rows, 3);
    assert.strictEqual(result.cols, 50);


    assert.strictEqual(result.combos[0].period, 10);
    assert.strictEqual(result.combos[1].period, 20);
    assert.strictEqual(result.combos[2].period, 30);
});

test('LinReg batch warmup consistency', () => {

    const data = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        data[i] = Math.random() * 100;
    }

    const result = wasm.linreg_batch(data, {
        period_range: [5, 15, 5]
    });

    const periods = [5, 10, 15];
    for (let row = 0; row < periods.length; row++) {
        const period = periods[row];
        const rowStart = row * 100;
        const rowData = result.values.slice(rowStart, rowStart + 100);


        const expectedWarmup = period - 1;
        for (let i = 0; i < expectedWarmup; i++) {
            assert(isNaN(rowData[i]), `Row ${row}: Expected NaN at index ${i}`);
        }


        for (let i = expectedWarmup; i < rowData.length; i++) {
            assert(!isNaN(rowData[i]), `Row ${row}: Unexpected NaN at index ${i}`);
        }
    }
});

File: tests/python/test_linreg.py
"""
Python binding tests for LinReg indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


try:
    from my_project import (
        linreg,
        linreg_batch,
        LinRegStream
    )
except ImportError:
    pytest.skip("LinReg module not available - run 'maturin develop' first", allow_module_level=True)


def test_linreg_partial_params():
    """Test LinReg with default parameters - mirrors check_linreg_partial_params"""
    data = load_test_data()
    close = np.array(data['close'], dtype=np.float64)


    result = linreg(close, 14)
    assert len(result) == len(close)


def test_linreg_accuracy():
    """Test LinReg matches expected values from Rust tests - mirrors check_linreg_accuracy"""
    data = load_test_data()
    close = np.array(data['close'], dtype=np.float64)
    expected = EXPECTED_OUTPUTS['linreg']


    result = linreg(close, expected['default_params']['period'])


    assert_close(
        result[-5:],
        expected['last_5_values'],
        atol=1e-1,
        msg="LinReg last 5 values mismatch"
    )


    compare_with_rust('linreg', result, 'close', expected['default_params'])


def test_linreg_zero_period():
    """Test LinReg fails with zero period - mirrors check_linreg_zero_period"""
    input_data = np.array([10.0, 20.0, 30.0], dtype=np.float64)

    with pytest.raises(ValueError, match="Invalid period"):
        linreg(input_data, 0)


def test_linreg_period_exceeds_length():
    """Test LinReg fails with period exceeding data length - mirrors check_linreg_period_exceeds_length"""
    data_small = np.array([10.0, 20.0, 30.0], dtype=np.float64)

    with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
        linreg(data_small, 10)


def test_linreg_very_small_dataset():
    """Test LinReg fails with insufficient data - mirrors check_linreg_very_small_dataset"""
    single_point = np.array([42.0], dtype=np.float64)

    with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
        linreg(single_point, 14)


def test_linreg_empty_input():
    """Test LinReg with empty input"""
    data_empty = np.array([], dtype=np.float64)

    with pytest.raises(ValueError, match="All values are NaN"):
        linreg(data_empty, 14)


def test_linreg_all_nan():
    """Test LinReg with all NaN input"""
    data = np.array([np.nan, np.nan, np.nan, np.nan, np.nan], dtype=np.float64)

    with pytest.raises(ValueError, match="All values are NaN"):
        linreg(data, 3)


def test_linreg_reinput():
    """Test LinReg with re-input of LinReg result - mirrors check_linreg_reinput"""
    data = load_test_data()
    close = np.array(data['close'], dtype=np.float64)


    first_result = linreg(close, 14)


    second_result = linreg(first_result, 10)

    assert len(second_result) == len(first_result)





    for i in range(24, len(second_result)):
        assert np.isfinite(second_result[i]), f"Unexpected NaN at index {i}"


def test_linreg_nan_handling():
    """Test LinReg handling of NaN values - mirrors check_linreg_nan_handling"""
    data = load_test_data()
    close = np.array(data['close'], dtype=np.float64)
    expected = EXPECTED_OUTPUTS['linreg']


    result = linreg(close, expected['default_params']['period'])

    assert len(result) == len(close)


    first_valid = next((i for i, x in enumerate(result) if not np.isnan(x)), None)
    assert first_valid == expected['warmup_period'], f"Expected warmup at {expected['warmup_period']}, got {first_valid}"



    if len(result) > 240:
        for i in range(240, len(result)):
            assert not np.isnan(result[i]), f"Found unexpected NaN at index {i}"


def test_linreg_batch():
    """Test LinReg batch computation"""
    data = load_test_data()
    close = np.array(data['close'], dtype=np.float64)


    period_start = 10
    period_end = 40
    period_step = 10

    result = linreg_batch(close, (period_start, period_end, period_step))

    assert 'values' in result
    assert 'periods' in result


    assert len(result['periods']) == 4
    assert np.array_equal(result['periods'], [10, 20, 30, 40])


    assert result['values'].shape == (4, len(close))


    individual_result = linreg(close, 10)
    batch_first_row = result['values'][0]
    assert_close(batch_first_row, individual_result, atol=1e-9, msg="First combination mismatch")


def test_linreg_batch_multiple_periods():
    """Test LinReg batch computation with multiple periods"""
    data = load_test_data()
    close = np.array(data['close'], dtype=np.float64)


    period_start = 20
    period_end = 60
    period_step = 20

    result = linreg_batch(close, (period_start, period_end, period_step))


    assert 'values' in result
    assert 'periods' in result
    assert np.array_equal(result['periods'], [20, 40, 60])


    assert result['values'].shape == (3, len(close))


    individual_result = linreg(close, 40)
    batch_second_row = result['values'][1]
    assert_close(batch_second_row, individual_result, atol=1e-9, msg="Second combination mismatch")


def test_linreg_batch_single_period():
    """Test LinReg batch computation with single period"""
    data = load_test_data()
    close = np.array(data['close'], dtype=np.float64)


    result = linreg_batch(close, (30, 30, 0))


    assert 'values' in result
    assert 'periods' in result
    assert np.array_equal(result['periods'], [30])


    assert result['values'].shape == (1, len(close))


    individual_result = linreg(close, 30)
    assert_close(result['values'][0], individual_result, atol=1e-9, msg="Single period mismatch")


def test_linreg_stream():
    """Test LinReg streaming interface - mirrors check_linreg_streaming"""
    data = load_test_data()
    close = data['close']
    period = 14


    close_array = np.array(close, dtype=np.float64)
    batch_result = linreg(close_array, period)


    stream = LinRegStream(period)
    stream_results = []

    for price in close:
        result = stream.update(price)

        stream_results.append(result if result is not None else np.nan)


    assert len(batch_result) == len(stream_results)

    for i in range(len(batch_result)):

        if np.isnan(batch_result[i]) and np.isnan(stream_results[i]):
            continue

        assert_close(stream_results[i], batch_result[i], atol=1e-6,
                    msg=f"Streaming mismatch at index {i}")


def test_linreg_different_periods():
    """Test LinReg with various period values"""
    data = load_test_data()
    close = np.array(data['close'], dtype=np.float64)


    for period in [5, 10, 20, 50]:
        result = linreg(close, period)
        assert len(result) == len(close)


        valid_count = np.sum(np.isfinite(result[period:]))
        assert valid_count > len(close) - period - 5, f"Too many NaN values for period={period}"


def test_linreg_batch_performance():
    """Test that batch computation works correctly (performance is secondary)"""
    data = load_test_data()
    close = np.array(data['close'][:1000], dtype=np.float64)


    result = linreg_batch(close, (10, 50, 10))
    assert len(result['periods']) == 5


    for i, period in enumerate(result['periods']):
        individual_result = linreg(close, period)
        batch_row = result['values'][i]
        assert_close(batch_row, individual_result, atol=1e-9,
                          msg=f"Batch mismatch for period={period}")


def test_linreg_edge_cases():
    """Test LinReg with edge case inputs"""

    data = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0], dtype=np.float64)
    result = linreg(data, 3)
    assert len(result) == len(data)



    assert_close(result[-1], 10.0, atol=1e-9, msg="Perfect linear regression failed")


    data = np.array([1.0, np.nan, 3.0, np.nan, 5.0, np.nan, 7.0, np.nan], dtype=np.float64)
    result = linreg(data, 2)
    assert len(result) == len(data)


    data = np.array([5.0] * 20, dtype=np.float64)
    result = linreg(data, 5)
    assert len(result) == len(data)

    for i in range(5, len(result)):
        assert_close(result[i], 5.0, atol=1e-9, msg=f"Constant prediction failed at index {i}")


def test_linreg_single_value():
    """Test LinReg with single value input"""
    data = np.array([42.0], dtype=np.float64)


    result = linreg(data, 1)
    assert len(result) == 1
    assert np.isnan(result[0])


def test_linreg_two_values():
    """Test LinReg with two values input"""
    data = np.array([1.0, 2.0], dtype=np.float64)



    result = linreg(data, 1)
    assert len(result) == 2
    assert np.isnan(result[0])
    assert np.isnan(result[1])


    result = linreg(data, 2)
    assert len(result) == 2
    assert np.isnan(result[0])
    assert_close(result[1], 2.0, atol=1e-9, msg="Two-value prediction failed")


def test_linreg_warmup_period():
    """Test that warmup period is correctly calculated (first + period - 1)"""
    data = load_test_data()
    close = np.array(data['close'][:50], dtype=np.float64)

    test_cases = [
        (5, 4),
        (10, 9),
        (20, 19),
        (30, 29),
    ]

    for period, expected_warmup in test_cases:
        result = linreg(close, period)


        for i in range(expected_warmup):
            assert np.isnan(result[i]), f"Expected NaN at index {i} for period={period}"


        if expected_warmup < len(result):
            assert not np.isnan(result[expected_warmup]), \
                f"Expected valid value at index {expected_warmup} for period={period}"


def test_linreg_consistency():
    """Test that LinReg produces consistent results across multiple calls"""
    data = load_test_data()
    close = np.array(data['close'][:100], dtype=np.float64)

    result1 = linreg(close, 14)
    result2 = linreg(close, 14)

    assert_close(result1, result2, atol=1e-15, msg="LinReg results not consistent")


def test_linreg_step_precision():
    """Test batch with very small step sizes"""
    data = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], dtype=np.float64)

    result = linreg_batch(data, (2, 4, 1))

    assert np.array_equal(result['periods'], [2, 3, 4])
    assert result['values'].shape == (3, len(data))


def test_linreg_slope_calculation():
    """Test LinReg slope calculation with known data"""

    data = np.array([2.0, 4.0, 6.0, 8.0, 10.0, 12.0, 14.0, 16.0], dtype=np.float64)
    result = linreg(data, 4)



    assert_close(result[-1], 16.0, atol=1e-9, msg="Slope calculation failed")


def test_linreg_batch_error_handling():
    """Test LinReg batch error handling"""

    empty = np.array([], dtype=np.float64)
    with pytest.raises(ValueError, match="Input data slice is empty|All values are NaN"):
        linreg_batch(empty, (10, 20, 10))


    all_nan = np.full(100, np.nan, dtype=np.float64)
    with pytest.raises(ValueError, match="All values are NaN"):
        linreg_batch(all_nan, (10, 20, 10))


    small_data = np.array([1.0, 2.0, 3.0], dtype=np.float64)
    with pytest.raises(ValueError, match="Not enough valid data"):
        linreg_batch(small_data, (5, 10, 5))


def test_linreg_batch_metadata_consistency():
    """Test that batch metadata is consistent with results"""
    data = load_test_data()
    close = np.array(data['close'][:100], dtype=np.float64)


    test_cases = [
        (5, 15, 5),
        (10, 10, 0),
        (20, 40, 10),
    ]

    for start, end, step in test_cases:
        result = linreg_batch(close, (start, end, step))


        if step == 0 or start == end:
            expected_periods = [start]
        else:
            expected_periods = list(range(start, end + 1, step))

        assert np.array_equal(result['periods'], expected_periods)
        assert result['values'].shape[0] == len(expected_periods)
        assert result['values'].shape[1] == len(close)


def test_linreg_zero_copy_verification():
    """Verify LinReg uses zero-copy operations"""

    data = load_test_data()
    close = np.array(data['close'][:100], dtype=np.float64)


    result = linreg(close, 14)
    assert len(result) == len(close)


    batch_result = linreg_batch(close, (10, 30, 10))
    assert batch_result['values'].shape == (3, len(close))


def test_linreg_batch_warmup_consistency():
    """Test that batch warmup periods are consistent"""
    data = np.random.randn(50).astype(np.float64)

    result = linreg_batch(data, (5, 15, 5))


    for i, period in enumerate(result['periods']):
        row = result['values'][i]

        assert np.all(np.isnan(row[:period-1])), f"Expected NaN warmup for period {period}"

        assert np.all(~np.isnan(row[period-1:])), f"Expected values after warmup for period {period}"


def test_linreg_streaming():
    """Test LinReg streaming matches batch calculation - mirrors check_linreg_streaming"""
    data = load_test_data()
    close = np.array(data['close'][:100], dtype=np.float64)
    expected = EXPECTED_OUTPUTS['linreg']
    period = expected['default_params']['period']


    batch_result = linreg(close, period)


    stream = LinRegStream(period)
    stream_values = []

    for price in close:
        result = stream.update(price)
        stream_values.append(result if result is not None else np.nan)

    stream_values = np.array(stream_values)


    assert len(batch_result) == len(stream_values)


    for i, (b, s) in enumerate(zip(batch_result, stream_values)):
        if np.isnan(b) and np.isnan(s):
            continue
        assert_close(b, s, rtol=1e-9, atol=1e-9,
                    msg=f"LinReg streaming mismatch at index {i}")


def test_linreg_stream_error_handling():
    """Test LinReg stream error handling"""

    with pytest.raises(ValueError, match="LinRegStream error|Invalid period"):
        LinRegStream(0)


if __name__ == "__main__":

    print("Testing LinReg module...")
    test_linreg_accuracy()
    print("LinReg tests passed!")